crates/bevy_pbr/src/render/skin.rs

```diff
@@ -17,32 +17,118 @@ use bevy_utils::EntityHashMap;
 pub const MAX_JOINTS: usize = 256;
 
 #[derive(Component)]
-pub struct SkinIndex {
-    pub index: u32,
-}
-
-impl SkinIndex {
-    /// Index to be in address space based on [`SkinUniform`] size.
-    const fn new(start: usize) -> Self {
-        SkinIndex {
-            index: (start * std::mem::size_of::<Mat4>()) as u32,
-        }
-    }
+pub enum SkinIndex {
+    Index(u32),
+    DynamicOffset(u32),
 }
 
 #[derive(Default, Resource, Deref, DerefMut)]
 pub struct SkinIndices(EntityHashMap<Entity, SkinIndex>);
 
 // Notes on implementation: see comment on top of the `extract_skins` system.
 #[derive(Resource)]
-pub struct SkinUniform {
-    pub buffer: BufferVec<Mat4>,
+pub enum SkinUniform {
+    Uniform(BufferVec<Mat4>),
+    Storage(StorageBuffer<Vec<Mat4>>),
```

Comment:
Have you considered defining this struct as a standalone type in `bevy_render`? Similarly to `GpuArrayBuffer`? Even if it is only used here, it helps separating concerns. It will also make it easier to remediate conflicts with #9002 (independently of who has to fix the merge conflicts)

Reply:
The problem is that it's very specific to Mat4 at the moment. Ideally it feels like this should be a different 'mode' of GpuArrayBuffer. One mode for batching as many instances of data into one binding, and one where you have a known max binding size that you use for a uniform buffer, but you may not use it all so you align to the next dynamic offset binding as needed and make sure the last dynamic offset has a full binding of data after it. That requires more thought to implement.

Reply:
And if it weren't specific to Mat4, we could use an affine matrix instead and save space.

Reply:
I figured out how to do this, and would like to. But I think it has to wait until after 0.12. I'll move this to 0.13.